inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/activation_ref.cl

```diff
@@ -33,43 +33,53 @@ KERNEL(activation)(
-    const unsigned batch = 0;
-#else
-    const unsigned feature = (uint)get_global_id(2) % OUTPUT_FEATURE_NUM;
-    const unsigned batch = (uint)get_global_id(2) / OUTPUT_FEATURE_NUM;
-#endif
-#else
-#if defined OUTPUT_LAYOUT_YXFB || defined OUTPUT_LAYOUT_B_FS_YX_FSV16
-    const unsigned x = (uint)get_global_id(1);
-    const unsigned y = (uint)get_global_id(2);
-#define z 0
-#if OUTPUT_BATCH_NUM == 1
-    const unsigned feature = (uint)get_global_id(0);
-    const unsigned batch = 0;
-#else
-    const unsigned feature = (uint)get_global_id(0) % OUTPUT_FEATURE_NUM;
-    const unsigned batch = (uint)get_global_id(0) / OUTPUT_FEATURE_NUM;
-#endif
-#else
-#define z 0
-    const unsigned x = (uint)get_global_id(0);
-    const unsigned y = (uint)get_global_id(1);
-#if OUTPUT_BATCH_NUM == 1
-    const unsigned feature = (uint)get_global_id(2);
-    const unsigned batch = 0;
+    #if OUTPUT_BATCH_NUM == 1
+        const unsigned feature = (uint)get_global_id(2);
+        const unsigned batch = 0;
+    #else
+        const unsigned feature = (uint)get_global_id(2) % OUTPUT_FEATURE_NUM;
+        const unsigned batch = (uint)get_global_id(2) / OUTPUT_FEATURE_NUM;
+    #endif
 #else
-    const unsigned feature = (uint)get_global_id(2) % OUTPUT_FEATURE_NUM;
-    const unsigned batch = (uint)get_global_id(2) / OUTPUT_FEATURE_NUM;
-#endif
-#endif
+    #if defined OUTPUT_LAYOUT_YXFB || defined OUTPUT_LAYOUT_B_FS_YX_FSV16 || defined OUTPUT_LAYOUT_B_FS_YX_FSV32
+        const unsigned x = (uint)get_global_id(1);
+        const unsigned y = (uint)get_global_id(2);
+        #define z 0
+        #if OUTPUT_BATCH_NUM == 1
+            const unsigned feature = (uint)get_global_id(0);
+            const unsigned batch = 0;
+        #else
+            const unsigned feature = (uint)get_global_id(0) % OUTPUT_FEATURE_NUM;
+            const unsigned batch = (uint)get_global_id(0) / OUTPUT_FEATURE_NUM;
+        #endif
+    #elif defined OUTPUT_LAYOUT_BS_FS_YX_BSV32_FSV32 || defined OUTPUT_LAYOUT_BS_FS_YX_BSV32_FSV16
+        const unsigned x = (uint)get_global_id(0) % OUTPUT_SIZE_X;
+        const unsigned y = (uint)get_global_id(0) / OUTPUT_SIZE_X;
+        const unsigned feature = (uint)get_global_id(1);
+        const unsigned batch = (uint)get_global_id(2);
+    #else
+        #define z 0
+            const unsigned x = (uint)get_global_id(0);
+            const unsigned y = (uint)get_global_id(1);
+        #if OUTPUT_BATCH_NUM == 1
+            const unsigned feature = (uint)get_global_id(2);
+            const unsigned batch = 0;
+        #else
+            const unsigned feature = (uint)get_global_id(2) % OUTPUT_FEATURE_NUM;
+            const unsigned batch = (uint)get_global_id(2) / OUTPUT_FEATURE_NUM;
+        #endif
+    #endif
 #endif
 
-#if defined(OUTPUT_LAYOUT_B_FS_YX_FSV16) && OUTPUT_FEATURE_NUM % 16 != 0
+#if (defined(OUTPUT_LAYOUT_B_FS_YX_FSV16) || defined(OUTPUT_LAYOUT_B_FS_YX_FSV32)) \
+    && OUTPUT_FEATURE_NUM % 16 != 0
```

Comment:
What if `OUTPUT_FEATURE_NUM == 48` and `OUTPUT_LAYOUT_B_FS_YX_FSV32`?
Then I think this `#if` will be false, but we need to check `feature >= OUTPUT_FEATURE_NUM`.

Reply:
LWS for feature is aligned by 16 regardless of the output layout format. So, it deosn't need to check OUTPUT_FEATURE_NUM can be divded by 32 if the output layout is OUTPUT_LAYOUT_B_FS_YX_FSV32.

Reply:
// GWS.feature and GWS.batch is aligned to 16. Otherwise, there are some idling WIs.

Reply:
It looks like that all is OK with this code and there are no idling WIs and related perf problems. 

If (output_feature % 16 != 0) then (output_feature % 32 != 0) as well. 
And we really needed feature >= OUTPUT_FEATURE_NUM comparison. 
And the same for double blocked format bsv_fsv.

@isanghao take a look again please, the condition under #ifdef only regulates we need or not extra checkings for feature / batch range limitations. But using 16 instead of 32 can really simplifies the code and avoids duplications.

Reply:
@lznamens  Yeah, Jade and I aligned on this offline. You are right :)